### ISSUE TITLE
Interoperability: Clients running on different MP platforms need to know how to start and end and LRA

### ISSUE THREAD

**Issue by mmusgrov** [eclipse/microprofile-lra#62]
In MP environments running different MP implementations a client running on one platform must be able to start and end an LRA implemented by another vendor.

This requirement was briefly discussed in issue #17 .

The specification should:
- define using MP Configuration how the client can locate a REST endpoint where LRAs can be started and ended.
- define the REST payload and verbs (PUT, POST etc) for the requests to start and end LRAs

**Comment by xstefank** [comment]
As it was discussed on the hangout this week, here are starting proposed REST definitions:

start:
`POST /start`
parameters: 
*`ClientID`: query, optional, String
** a unique identity of a client starting LRA 
* TimeLimitL query, optional, int64
            "in": "query",
            "description": "Specifies the maximum time in milli seconds that the LRA will exist for.\nIf the LRA is terminated because of a timeout, the LRA URL is deleted.\nAll further invocations on the URL will return 404.\nThe invoker can assume this was equivalent to a compensate operation.",
            "required": false,
            "type": "integer",
            "default": 0,
            "format": "int64"
          },
          {
            "name": "ParentLRA",
            "in": "query",
            "description": "The enclosing LRA if this new LRA is nested",
            "required": false,
            "type": "string"
          },
          {
            "name": "Long-Running-Action",
            "in": "header",
            "required": false,
            "type": "string"
          }
        ],
        "responses": {
          "200": {
            "description": "The request was successful and the response body contains the id of the new LRA"
          },
          "500": {
            "description": "A new LRA could not be started"
          }
        }
      }

**Comment by xstefank** [comment]
As it was discussed on the hangout this week, here are starting proposed REST definitions:

**start**:
`POST /start`

parameters:
- `ClientID`: query, optional, String
  - a unique identity of a client starting LRA
- `TimeLimit`: query, optional, int64
  - Specifies the maximum time in milliseconds that the LRA will exist for
- `ParentLRA`: query, optional, String
  - The enclosing LRA if this new LRA is nested

responses: 
`200`: The request was successful and the response body contains the id of the new LRA
`500`: A new LRA could not be started, response body may optionally specify reason

Produces `text/plain`

**close**:
`PUT {lraId}/close`

paramerters:
- `lraId`: path, required, String
  - The unique identifier of the LRA

responses:
`200`: The complete message was sent to all participants
`404`: This LRA presumably does not exit

**cancel**:
`PUT {lraId}/cancel`

paramerters:
- `lraId`: path, required, String
  - The unique identifier of the LRA

responses:
`200`: The compensate message was sent to all participants
`404`: This LRA presumably does not exit

**Comment by mmusgrov** [comment]
We need to follow REST HATEOS principles, ie the POST request should return the newly created resource in the Location header and a link header containing three URLs that can be used to close, cancel and join the LRA (with relation types close , cancel and join respectively).

**Comment by mmusgrov** [comment]
After thinking further on your proposal and on what I said in my last comment, I have two questions/observations:

1) With a RESTful API POST creates new resources and the payload optionally specifies it's initial state (the list of participants is part of the state of an LRA so passing along an initial participant makes good sense).
Question: But is it good REST design to ask the server to infer the initial state via query params and link headers or should we require that JSON should be used?

2) The response to the POST should return a link to the new resource (in the Location header). But in the case of an LRA resource the cancel/close/join are really part of the state of the LRA (they are not resources like I implied). So I would suggest that to cancel an LRA the client should issue a PUT request to the LRA resource and the payload should set the state of the LRA to `cancelled`. The server would then attempt to transition the LRA to the `cancelled` state. Similarly for close.

**Comment by mmusgrov** [comment]
Also note that from my point of view this work should cover starting, closing and cancelling LRAs but should also include how to *join an LRA*: @rdebusscher may we move this issue back to #17 since starting, joining and ending are all tied to each other, ie starting an LRA should return something that participants can use to join with the new LRA?

**Comment by ochaloup** [comment]
I like the Mike's point about the JSON as configuration options. In addition what was already said as benefits it seems to me kind of "cleaner" design as I think JSON is more suitable for `POST` request data.

one think I don't understand now is the second point of Mike's comment

>     2. The response to the POST should return a link to the new resource (in the Location header). But in the case of an LRA resource the cancel/close/join are really part of the state of the LRA (they are not resources like I implied). So I would suggest that to cancel an LRA the client should issue a PUT request to the LRA resource and the payload should set the state of the LRA to `cancelled`. The server would then attempt to transition the LRA to the `cancelled` state. Similarly for close.

Does the commend proposes there will be only one URL from the `start` call and the cancel would not be called to the `PUT {lraId}/cancel` but to `PUT {lraId}` with body `cancelled`? Or this is thought differently?

**Comment by mmusgrov** [comment]
Ondra you are right. Starting an LRA will create a new LRA resource {lraId}. To close or cancel the LRA you would send a PUT request to {lraId} and the body of the request would contain some JSON that contains the desired resource representation that corresponds to the closed state:

```json
{
 ...
  "state": "closed"
  ...
}
```

**Comment by rdebusscher** [comment]
@mmusgrov Yes, we can merge this with #17. I don't know anymore why I thought issue #17 was about the integration of different coordinators.

**Comment by mmusgrov** [comment]
The [decision made at the last hangout](https://docs.google.com/document/d/1HBPiXpsXcQp9T8JeoT-IDs-EblZX1SxuTV5qjUPzwrM/edit?ts=5bfc2daa#bookmark=id.75jxt3khvwgf) was to defer this topic to milestone 1.2
